ddtrace/internal/gitmetadata.py

```diff
@@ -0,0 +1,126 @@
+
+
+def _get_tags_from_env(config):
+    # type: (GitMetadataConfig) -> typing.Tuple[str, str]
+    """
+    Get git metadata from environment variables.
+    Returns tuple (repository_url, commit_sha)
+    """
+    repository_url = config.repository_url
+    commit_sha = config.commit_sha
+
+    tags = formats.parse_tags_str(config.tags)
+    if repository_url == "":
+        repository_url = tags.get(REPOSITORY_URL, "")
+    if commit_sha == "":
+        commit_sha = tags.get(COMMIT_SHA, "")
+
+    return repository_url, commit_sha
+
+
+def _get_tags_from_package(config):
+    # type: (GitMetadataConfig) -> typing.Tuple[str, str]
+    """
+    Extracts git metadata from python package's medatada field Project-URL:
+    e.g: Project-URL: source_code_link, https://github.com/user/repo#gitcommitsha&someoptions
+    Returns tuple (repository_url, commit_sha)
+    """
+    if config.main_package == "":
+        return "", ""
+    try:
+        try:
+            import importlib.metadata as importlib_metadata
+        except ImportError:
+            import importlib_metadata  # type: ignore[no-redef]
+
+        source_code_link = ""
+        for val in importlib_metadata.metadata(config.main_package).get_all("Project-URL"):
+            capt_val = val.split(", ")
+            if capt_val[0] == "source_code_link":
+                source_code_link = capt_val[1].strip()
+                break
+
+        if source_code_link != "" and "#" in source_code_link:
+            repository_url, commit_sha = source_code_link.split("#")
+            commit_sha = commit_sha.split("&")[0]
+            return repository_url, commit_sha
+        return "", ""
+    except importlib_metadata.PackageNotFoundError:
+        return "", ""
+
+
+def get_git_tags():
+    # type: () -> typing.Tuple[str, str]
+    """
+    Returns git metadata tags tuple (repository_url, commit_sha)
+    """
+    try:
+        global _GITMETADATA_TAGS
+        if _GITMETADATA_TAGS is not None:
+            return _GITMETADATA_TAGS
+
+        config = GitMetadataConfig()
+
+        if config.enabled:
+            repository_url, commit_sha = _get_tags_from_env(config)
+            log.debug("git tags from env: %s %s", repository_url, commit_sha)
+            if repository_url == "" or commit_sha == "":
+                pkg_repository_url, pkg_commit_sha = _get_tags_from_package(config)
+                log.debug("git tags from package: %s %s", pkg_repository_url, pkg_commit_sha)
+                if repository_url == "":
+                    repository_url = pkg_repository_url
+                if commit_sha == "":
+                    commit_sha = pkg_commit_sha
+
+            log.debug("git tags: %s %s", repository_url, commit_sha)
+            _GITMETADATA_TAGS = repository_url, commit_sha
+        else:
+            log.debug("git tags disabled")
+            _GITMETADATA_TAGS = ("", "")
+        return _GITMETADATA_TAGS
+    except Exception:
+        log.exception("get_git_tags failed")
```

Comment:
At general we don't expect errors, but if some malformed data comes, or something like it we would like to don't break, but inform that something strange happens.

Reply:
👍🏻 agree with the try/except, what I am curious about is the log message.

since an exception log will get shown to the user (likely not long after startup). my question was mostly, if they see that log message (with traceback) then is there anything they can do? is it clear the actions they should take?

which makes me wonder if the message itself should be more actionable like... suggest having them look at the specific env vars / link to product page on how the feature works / ask the to open a support ticket ?

or.... should we just make it a debug log? the flow would likely be, they try to get git metadata setup, it isn't working, they open a support case, we collect debug logs, and then we do the interpreting of the exception (not ideal flow for timeliness, but a valid one).

Reply:
Ok, agree with you, lets move it to debug